### ISSUE TITLE
Service contracts vs using methos from implementations

### ISSUE THREAD

**Issue by bartoszkubicki** [magento/magento2#21635]
### Summary (*)

Service contracts are really great, but there are a lot of situations when they are hard to follow. There are functions from implementations used in code or these functions are really useful and they tempt to use themselves (as we know what is the implementations and it is not likely that it will be changed). In effect, contract service are used, but in fact it impossible to change its implementations, as code would brake, because we rely on concrete implementations. Of course, we should not do it - but sometimes, we have to override some code, which actually is doing it this way. In adddition it makes testing harder, because we have to mock implementations instead of interfaces (unless we would do it, we would get warning from phpunit, that method doesn't exist or sth).
What we can do in such a situation - our class needs function from implementation, which is not available in interface? I guess we should require implementation only, not interface.

### Examples (*)

`ProductInterface` and `Product`. `Product` has method `getFinalPrice`. Of course if we take a look inside, maybe we can take `Magento\Catalog\Model\Product\Type\Price\Factory` and try to load proper `price model` for our product type (we have `getTypeId` function for our interface), but it is reinventing a wheel and what is more, `Product` method `getFinalPrice`, has some local caching, so its beneficial for performance, as for single cart reload this function is called many times. So if need `getFinalPrice` should I require in my class just `Product`? Seems to be the most reasonable, although it leads me out of service contract, and then tempts to increase of usage such a functions, so later refactor for replacing implementation with interface its even harder.

### Proposed solution

Code should be refactored. Some new interface should be introduced and public methods in implementations not present in interfaces should be deprecated and eventually removed. Although occurances of these problems are numerous, so here I am just asking, what to do, when we have conflict interface vs implementation.

**Comment by magento-engcom-team** [comment]
Hi @bartoszkubicki. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@bartoszkubicki do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by ghost** [comment]
Hi @bartoszkubicki  It will be processed faster if you move this to https://github.com/magento/community-features becouse it need to introduce a big  code changes. Thanks for collaboration.
